source/common/network/io_socket_handle_impl.cc

```diff
@@ -55,5 +61,56 @@ IoHandleCallUintResult IoSocketHandleImpl::close() {
 
 bool IoSocketHandleImpl::isOpen() const { return fd_ != -1; }
 
+IoHandleCallUintResult IoSocketHandleImpl::readv(uint64_t max_length, Buffer::RawSlice* slices,
+                                                 uint64_t num_slice) {
+  STACK_ARRAY(iov, iovec, num_slice);
+  uint64_t num_slices_to_read = 0;
+  uint64_t num_bytes_to_read = 0;
+  for (; num_slices_to_read < num_slice && num_bytes_to_read < max_length; num_slices_to_read++) {
+    iov[num_slices_to_read].iov_base = slices[num_slices_to_read].mem_;
+    const size_t slice_length = std::min(slices[num_slices_to_read].len_,
+                                         static_cast<size_t>(max_length - num_bytes_to_read));
+    iov[num_slices_to_read].iov_len = slice_length;
+    num_bytes_to_read += slice_length;
+  }
+  ASSERT(num_bytes_to_read <= max_length);
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  const Api::SysCallSizeResult result =
+      os_syscalls.readv(fd_, iov.begin(), static_cast<int>(num_slices_to_read));
```

Comment:
Would it be possible to call readv directly here? I ask, since for our Windows implementation, we'll be using _WSABUF / WSARecv instead of iovec / readv, and we don't to use platform-specific typedefs if at all possible (which we'd have to do to have a readv/writev implementation in OsSysCalls)

I think these tests: https://github.com/envoyproxy/envoy/blob/99696cda3336af26fe0b048d91e0d6eb279bb81c/test/common/buffer/owned_impl_test.cc#L115 would have to use a mocked IoSocketHandle though

Reply:
As you see, IoSocketHandleImpl::readv/writev now does more than calling a system api. If we mock out the IoHandle code, many of the existing unit tests won't be able to cover the actually IO logic.

One thing I don't really understand about the problem: I thought windows implementation will have its own IoHandle implementation,, why would using os_sys_call here matter at all to a totally different implementation?

Reply:
Unless I'm missing something, there aren't actually any unit tests that cover this IO logic. The only two test files that use mocked readv / writev are:
1. owned_impl_test.cc, which doesn't make any assertions on the arguments passed to readv/writev
2. proxy_protocol_test.cc, which just passes the arguments through to the actual syscall.

So I do not think that using a mocked IoSocketHandle would affect the current test coverage. I would also argue that the Buffer::OwnedImpl tests shouldn't "know" the implementation of IoSocketHandle 

If we do want to add tests that cover how IoSocketHandleImpl::readv/writev implement the IO logic, they should probably test the class directly

As to why (from a Windows perspective), we care about this separate implementation: we would like to remove writev and readv from the OsSysCalls interface, since their function signatures cannot be compiled on Windows (iovec is not a type on Windows).

If we did need to keep writev + readv in the OsSysCalls interface, we would probably just end up with something like

```
class OsSysCalls {
#ifdef WIN32
// Windows system calls
#else 
// POSIX system calls
#endif
}
````

which isn't the worst thing in the world, but it's not ideal.

Reply:
> Unless I'm missing something, there aren't actually any unit tests that cover this IO logic. The only two test files that use mocked readv / writev are:
> 
> 1. owned_impl_test.cc, which doesn't make any assertions on the arguments passed to readv/writev
> 2. proxy_protocol_test.cc, which just passes the arguments through to the actual syscall.
> 
> So I do not think that using a mocked IoSocketHandle would affect the current test coverage. I 

IoSocketHandleImple::readv/writev is mostly a partial copy from Buffer::OwnedImpl::read/write. owned_impl_test.cc should have had coverage on them. If we mocked out IoSocketHandle, we will lose coverage on the IoSocketHandleImpl::readv/writev. It is true that we could test this class independently. But this PR is just a refactory, I don't want to add more complexity to it. Plus I could see more than just these 2 tests are using MockOsSysCalls, i.e. proxy_protocol_test.cc which using mocked IoSocketHandleImpl will require some plumbing through connection and buffer.

Do you think it's okay to keep the usage of OsSysCalls as is for now and add a TODO for future change if needed.

And just FYI, I think the get/setsockopt will use OsSysCalls API too.

Reply:
Right, there are other tests that use MockOsSysCalls, but owned_impl_test.cc and proxy_protocol_test.cc are the only 2 that actually care about readv / writev. And since neither test makes any assertions on the arguments to readv / writev, I don't think we actually have any coverage on the bits of IoSocketHandleImpl::readv/writev that were copied from Buffer::OwnedImpl::read/write.

Anyways, we do not need to solve this right now, we will come back to it when we start PR'ing the Windows implementation of IoSocketHandle.